src/extension/rpc.ts

```diff
@@ -0,0 +1,182 @@
+import type { NoInfer, SafeAny } from "../types";
+import type { MessageAdapter } from "./messageAdapters";
+import type {
+  RPCMessage,
+  RPCRequestMessage,
+  RPCResponseMessage,
+} from "./messages";
+import {
+  MessageType,
+  isRPCMessage,
+  isRPCRequestMessage,
+  isRPCResponseMessage,
+} from "./messages";
+
+type MessageCollection = Record<string, (...parameters: SafeAny[]) => SafeAny>;
+
+export interface RpcClient<Messages extends MessageCollection> {
+  withTimeout: (timeoutMs: number) => RpcClient<Messages>;
+  request: <TName extends keyof Messages & string>(
+    name: TName,
+    ...params: Parameters<Messages[TName]>
+  ) => Promise<Awaited<ReturnType<Messages[TName]>>>;
+}
+
+const DEFAULT_TIMEOUT = 30_000;
+
+export function createRpcClient<Messages extends MessageCollection>(
+  adapter: MessageAdapter<RPCRequestMessage>
+): RpcClient<Messages> {
+  let timeout = DEFAULT_TIMEOUT;
+
+  const client: RpcClient<Messages> = {
+    withTimeout: (timeoutMs) => {
+      timeout = timeoutMs;
+      return client;
+    },
+    request: (name, ...params) => {
+      const configuredTimeout = timeout;
+      timeout = DEFAULT_TIMEOUT;
+
+      return new Promise<SafeAny>((resolve, reject) => {
+        const id = createId();
+
+        const timeout = setTimeout(() => {
+          removeListener();
+          reject(new Error("Timeout waiting for message"));
+        }, configuredTimeout);
+
```

Comment:
I'm not perfectly happy with the `timeout` variable, as this expects `request` to be called immediately after `withTimeout`.

What do you think about this?


```suggestion
export interface RpcClient<Messages extends MessageCollection> {
  readonly timeout?: number;
  withTimeout: (
    this: RpcClient<Messages>,
    timeoutMs: number
  ) => RpcClient<Messages>;
  request: <TName extends keyof Messages & string>(
    this: RpcClient<Messages>,
    name: TName,
    ...params: Parameters<Messages[TName]>
  ) => Promise<Awaited<ReturnType<Messages[TName]>>>;
}

const DEFAULT_TIMEOUT = 30_000;

export function createRpcClient<Messages extends MessageCollection>(
  adapter: MessageAdapter<RPCRequestMessage>
): RpcClient<Messages> {
  const client: RpcClient<Messages> = {
    timeout: DEFAULT_TIMEOUT,
    withTimeout(timeoutMs) {
      return { ...this, timeout: timeoutMs };
    },
    request(name, ...params) {
      return new Promise<SafeAny>((resolve, reject) => {
        const id = createId();

        const timeout = setTimeout(() => {
          removeListener();
          reject(new Error("Timeout waiting for message"));
        }, this.timeout);

```

Reply:
I like it. I noodled on this idea, but after seeing it in the test, I'm sold. Thanks!

Reply:
[2b0f59a](https://github.com/apollographql/apollo-client-devtools/pull/1256/commits/2b0f59a83a3672f27b880d73312a8e84f14ad28e)